File: frontend/src/utils/connectionParser.ts
/**
 * URL Parser for Simplified Connectors
 * Converts connection URLs/strings into MindsDB-compatible configuration
 */

export interface ParsedConnection {
  connection_config: Record<string, any>;
  credentials: Record<string, any>;
}

export interface ConnectionParseError {
  success: false;
  error: string;
}

export interface ConnectionParseSuccess {
  success: true;
  data: ParsedConnection;
}

export type ConnectionParseResult = ConnectionParseSuccess | ConnectionParseError;

/**
 * Parse a connection URL/string into connection_config and credentials
 */
export function parseConnectionUrl(url: string, connectorType: string): ConnectionParseResult {
  try {
    switch (connectorType) {
      case 'mysql':
        return parseMySQLUrl(url);
      case 'postgresql':
        return parsePostgreSQLUrl(url);
      case 's3':
        return parseS3Url(url);
      case 'mongodb':
        return parseMongoDBUrl(url);
      case 'clickhouse':
        return parseClickHouseUrl(url);
      case 'api':
        return parseAPIUrl(url);
      default:
        return {
          success: false,
          error: `Unsupported connector type: ${connectorType}`
        };
    }
  } catch (error) {
    return {
      success: false,
      error: `Failed to parse connection URL: ${error instanceof Error ? error.message : 'Unknown error'}`
    };
  }
}

/**
 * Parse MySQL connection URL
 * Format: mysql://username:password@host:port/database?options
 */
function parseMySQLUrl(url: string): ConnectionParseResult {
  try {
    const parsed = new URL(url);
    
    if (parsed.protocol !== 'mysql:') {
      return { success: false, error: 'URL must start with mysql://' };
    }

    const host = parsed.hostname;
    const port = parsed.port ? parseInt(parsed.port) : 3306;
    const database = parsed.pathname.slice(1); // Remove leading slash
    const username = parsed.username;
    const password = parsed.password;

    if (!host || !database || !username) {
      return { 
        success: false, 
        error: 'MySQL URL must include host, database, and username' 
      };
    }

    return {
      success: true,
      data: {
        connection_config: {
          host,
          port,
          database
        },
        credentials: {
          user: username,
          password: password || ''
        }
      }
    };
  } catch (error) {
    return { 
      success: false, 
      error: 'Invalid MySQL URL format. Use: mysql://username:password@host:port/database' 
    };
  }
}

/**
 * Parse PostgreSQL connection URL
 * Format: postgresql://username:password@host:port/database?options
 */
function parsePostgreSQLUrl(url: string): ConnectionParseResult {
  try {
    const parsed = new URL(url);
    
    if (!['postgresql:', 'postgres:'].includes(parsed.protocol)) {
      return { success: false, error: 'URL must start with postgresql:// or postgres://' };
    }

    const host = parsed.hostname;
    const port = parsed.port ? parseInt(parsed.port) : 5432;
    const database = parsed.pathname.slice(1); // Remove leading slash
    const username = parsed.username;
    const password = parsed.password;

    if (!host || !database || !username) {
      return { 
        success: false, 
        error: 'PostgreSQL URL must include host, database, and username' 
      };
    }

    return {
      success: true,
      data: {
        connection_config: {
          host,
          port,
          database
        },
        credentials: {
          user: username,
          password: password || ''
        }
      }
    };
  } catch (error) {
    return { 
      success: false, 
      error: 'Invalid PostgreSQL URL format. Use: postgresql://username:password@host:port/database' 
    };
  }
}

/**
 * Parse S3 connection URL
 * Format: s3://access_key:secret_key@bucket_name/region
 */
function parseS3Url(url: string): ConnectionParseResult {
  try {
    const parsed = new URL(url);
    
    if (parsed.protocol !== 's3:') {
      return { success: false, error: 'URL must start with s3://' };
    }

    const bucketName = parsed.hostname;
    const region = parsed.pathname.slice(1) || 'us-east-1'; // Remove leading slash, default to us-east-1
    const accessKey = parsed.username;
    const secretKey = parsed.password;

    if (!bucketName || !accessKey || !secretKey) {
      return { 
        success: false, 
        error: 'S3 URL must include bucket name, access key, and secret key' 
      };
    }

    return {
      success: true,
      data: {
        connection_config: {
          bucket_name: bucketName,
          region
        },
        credentials: {
          aws_access_key_id: accessKey,
          aws_secret_access_key: secretKey
        }
      }
    };
  } catch (error) {
    return { 
      success: false, 
      error: 'Invalid S3 URL format. Use: s3://access_key:secret_key@bucket_name/region' 
    };
  }
}

/**
 * Parse MongoDB connection URL
 * Format: mongodb://username:password@host:port/database?options
 */
function parseMongoDBUrl(url: string): ConnectionParseResult {
  try {
    const parsed = new URL(url);
    
    if (parsed.protocol !== 'mongodb:') {
      return { success: false, error: 'URL must start with mongodb://' };
    }

    const host = parsed.hostname;
    const port = parsed.port ? parseInt(parsed.port) : 27017;
    const database = parsed.pathname.slice(1); // Remove leading slash
    const username = parsed.username;
    const password = parsed.password;

    if (!host || !database) {
      return { 
        success: false, 
        error: 'MongoDB URL must include host and database' 
      };
    }

    return {
      success: true,
      data: {
        connection_config: {
          host,
          port,
          database
        },
        credentials: {
          username: username || '',
          password: password || ''
        }
      }
    };
  } catch (error) {
    return { 
      success: false, 
      error: 'Invalid MongoDB URL format. Use: mongodb://username:password@host:port/database' 
    };
  }
}

/**
 * Parse ClickHouse connection URL
 * Format: clickhouse://username:password@host:port/database?options
 */
function parseClickHouseUrl(url: string): ConnectionParseResult {
  try {
    const parsed = new URL(url);
    
    if (parsed.protocol !== 'clickhouse:') {
      return { success: false, error: 'URL must start with clickhouse://' };
    }

    const host = parsed.hostname;
    const port = parsed.port ? parseInt(parsed.port) : 9000;
    const database = parsed.pathname.slice(1) || 'default'; // Remove leading slash, default to 'default'
    const username = parsed.username || 'default';
    const password = parsed.password;

    if (!host) {
      return { 
        success: false, 
        error: 'ClickHouse URL must include host' 
      };
    }

    return {
      success: true,
      data: {
        connection_config: {
          host,
          port,
          database
        },
        credentials: {
          user: username,
          password: password || ''
        }
      }
    };
  } catch (error) {
    return { 
      success: false, 
      error: 'Invalid ClickHouse URL format. Use: clickhouse://username:password@host:port/database' 
    };
  }
}

/**
 * Parse API connection URL
 * Format: https://api.example.com/endpoint?api_key=key&param=value
 */
function parseAPIUrl(url: string): ConnectionParseResult {
  try {
    const parsed = new URL(url);
    
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      return { success: false, error: 'API URL must start with http:// or https://' };
    }

    const baseUrl = `${parsed.protocol}//${parsed.host}`;
    const endpoint = parsed.pathname;
    const searchParams = new URLSearchParams(parsed.search);
    
    // Extract common API authentication parameters
    const apiKey = searchParams.get('api_key') || searchParams.get('apikey') || searchParams.get('key');
    const token = searchParams.get('token') || searchParams.get('access_token');
    const authHeader = searchParams.get('auth_header') || 'Authorization';

    // Remove auth params from the endpoint URL to keep it clean
    if (apiKey) {
      searchParams.delete('api_key');
      searchParams.delete('apikey');
      searchParams.delete('key');
    }
    if (token) {
      searchParams.delete('token');
      searchParams.delete('access_token');
    }

    const cleanEndpoint = endpoint + (searchParams.toString() ? `?${searchParams.toString()}` : '');

    return {
      success: true,
      data: {
        connection_config: {
          base_url: baseUrl,
          endpoint: cleanEndpoint,
          method: 'GET'
        },
        credentials: {
          api_key: apiKey || token || '',
          auth_header: authHeader
        }
      }
    };
  } catch (error) {
    return { 
      success: false, 
      error: 'Invalid API URL format. Use: https://api.example.com/endpoint?api_key=your_key' 
    };
  }
}

/**
 * Generate example URLs for each connector type
 */
export const CONNECTION_EXAMPLES = {
  mysql: 'mysql://user:password@localhost:3306/mydb',
  postgresql: 'postgresql://user:password@localhost:5432/postgres',
  s3: 's3://AKIA...:secret@my-bucket/us-east-1',
  mongodb: 'mongodb://user:password@localhost:27017/mydb',
  clickhouse: 'clickhouse://default:password@localhost:9000/default',
  api: 'https://api.example.com/users?api_key=abc123'
};

/**
 * Validate a connection URL format without parsing
 */
export function validateConnectionUrl(url: string, connectorType: string): { valid: boolean; error?: string } {
  const result = parseConnectionUrl(url, connectorType);
  return {
    valid: result.success,
    error: result.success ? undefined : result.error
  };
}

File: backend/app/utils/connection_parser.py
"""
Connection URL Parser for Backend
Converts connection URLs into MindsDB-compatible configuration
"""

from typing import Dict, Any, Tuple, Optional
from urllib.parse import urlparse, parse_qs
import re


class ConnectionParseError(Exception):
    """Exception raised when connection URL parsing fails"""
    pass


def parse_connection_url(url: str, connector_type: str) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """
    Parse a connection URL into connection_config and credentials
    
    Args:
        url: Connection URL string
        connector_type: Type of connector (mysql, postgresql, s3, etc.)
    
    Returns:
        Tuple of (connection_config, credentials)
    
    Raises:
        ConnectionParseError: If URL parsing fails
    """
    try:
        if connector_type == 'mysql':
            return parse_mysql_url(url)
        elif connector_type == 'postgresql':
            return parse_postgresql_url(url)
        elif connector_type == 's3':
            return parse_s3_url(url)
        elif connector_type == 'mongodb':
            return parse_mongodb_url(url)
        elif connector_type == 'clickhouse':
            return parse_clickhouse_url(url)
        elif connector_type == 'api':
            return parse_api_url(url)
        else:
            raise ConnectionParseError(f"Unsupported connector type: {connector_type}")
    except Exception as e:
        if isinstance(e, ConnectionParseError):
            raise
        raise ConnectionParseError(f"Failed to parse connection URL: {str(e)}")


def parse_mysql_url(url: str) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """Parse MySQL connection URL"""
    parsed = urlparse(url)
    
    if parsed.scheme != 'mysql':
        raise ConnectionParseError("URL must start with mysql://")
    
    if not all([parsed.hostname, parsed.path[1:], parsed.username]):
        raise ConnectionParseError("MySQL URL must include host, database, and username")
    
    # Apply SSL configuration for localhost development
    from app.core.config import settings
    
    port = parsed.port or 3306
    connection_config = {
        "host": parsed.hostname,
        "port": port,
        "database": parsed.path[1:]  # Remove leading slash
    }
    
    # Get SSL-aware configuration
    ssl_config = settings.get_ssl_config_for_connector(
        'mysql', parsed.hostname, port, connection_config
    )
    connection_config.update(ssl_config)
    
    credentials = {
        "user": parsed.username,
        "password": parsed.password or ""
    }
    
    return connection_config, credentials


def parse_postgresql_url(url: str) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """Parse PostgreSQL connection URL"""
    parsed = urlparse(url)
    
    if parsed.scheme not in ['postgresql', 'postgres']:
        raise ConnectionParseError("URL must start with postgresql:// or postgres://")
    
    if not all([parsed.hostname, parsed.path[1:], parsed.username]):
        raise ConnectionParseError("PostgreSQL URL must include host, database, and username")
    
    # Apply SSL configuration for localhost development
    from app.core.config import settings
    
    port = parsed.port or 5432
    connection_config = {
        "host": parsed.hostname,
        "port": port,
        "database": parsed.path[1:]  # Remove leading slash
    }
    
    # Get SSL-aware configuration
    ssl_config = settings.get_ssl_config_for_connector(
        'postgresql', parsed.hostname, port, connection_config
    )
    connection_config.update(ssl_config)
    
    credentials = {
        "user": parsed.username,
        "password": parsed.password or ""
    }
    
    return connection_config, credentials


def parse_s3_url(url: str) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """Parse S3 connection URL"""
    parsed = urlparse(url)
    
    if parsed.scheme != 's3':
        raise ConnectionParseError("URL must start with s3://")
    
    if not all([parsed.hostname, parsed.username, parsed.password]):
        raise ConnectionParseError("S3 URL must include bucket name, access key, and secret key")
    
    # Apply SSL configuration for development
    from app.core.config import settings
    
    # For S3, we'll use the bucket name as the host for SSL detection
    host = parsed.hostname
    
    connection_config = {
        "bucket_name": host,
        "region": parsed.path[1:] if parsed.path[1:] else "us-east-1"  # Remove leading slash, default region
    }
    
    # Get SSL-aware configuration
    ssl_config = settings.get_ssl_config_for_connector(
        's3', host, None, connection_config
    )
    connection_config.update(ssl_config)
    
    credentials = {
        "aws_access_key_id": parsed.username,
        "aws_secret_access_key": parsed.password
    }
    
    return connection_config, credentials


def parse_mongodb_url(url: str) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """Parse MongoDB connection URL"""
    parsed = urlparse(url)
    
    if parsed.scheme != 'mongodb':
        raise ConnectionParseError("URL must start with mongodb://")
    
    if not all([parsed.hostname, parsed.path[1:]]):
        raise ConnectionParseError("MongoDB URL must include host and database")
    
    # Apply SSL configuration for development
    from app.core.config import settings
    
    host = parsed.hostname
    port = parsed.port or 27017
    
    connection_config = {
        "host": host,
        "port": port,
        "database": parsed.path[1:]  # Remove leading slash
    }
    
    # Get SSL-aware configuration
    ssl_config = settings.get_ssl_config_for_connector(
        'mongodb', host, port, connection_config
    )
    connection_config.update(ssl_config)
    
    credentials = {
        "username": parsed.username or "",
        "password": parsed.password or ""
    }
    
    return connection_config, credentials


def parse_clickhouse_url(url: str) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """Parse ClickHouse connection URL"""
    parsed = urlparse(url)
    
    if parsed.scheme != 'clickhouse':
        raise ConnectionParseError("URL must start with clickhouse://")
    
    if not parsed.hostname:
        raise ConnectionParseError("ClickHouse URL must include host")
    
    # Apply SSL configuration for development
    from app.core.config import settings
    
    host = parsed.hostname
    port = parsed.port or 9000
    
    connection_config = {
        "host": host,
        "port": port,
        "database": parsed.path[1:] if parsed.path[1:] else "default"  # Remove leading slash, default database
    }
    
    # Get SSL-aware configuration
    ssl_config = settings.get_ssl_config_for_connector(
        'clickhouse', host, port, connection_config
    )
    connection_config.update(ssl_config)
    
    credentials = {
        "user": parsed.username or "default",
        "password": parsed.password or ""
    }
    
    return connection_config, credentials


def parse_api_url(url: str) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """Parse API connection URL"""
    parsed = urlparse(url)
    
    if parsed.scheme not in ['http', 'https']:
        raise ConnectionParseError("API URL must start with http:// or https://")
    
    # Apply SSL configuration for development
    from app.core.config import settings
    
    host = parsed.hostname
    port = parsed.port
    
    base_url = f"{parsed.scheme}://{parsed.netloc}"
    query_params = parse_qs(parsed.query)
    
    # Get SSL-aware configuration
    ssl_config = settings.get_ssl_config_for_connector(
        'api', host, port, {'base_url': base_url}
    )
    
    # Use the potentially modified base_url from SSL configuration
    base_url = ssl_config.get('base_url', base_url)
    
    # Extract common API authentication parameters
    api_key = None
    auth_header = "Authorization"
    
    # Look for various API key parameter names
    for key_name in ['api_key', 'apikey', 'key', 'token', 'access_token']:
        if key_name in query_params:
            api_key = query_params[key_name][0]  # Take first value
            break
    
    # Remove auth params from the endpoint URL to keep it clean
    clean_query_params = {k: v for k, v in query_params.items() 
                         if k not in ['api_key', 'apikey', 'key', 'token', 'access_token']}
    
    # Reconstruct clean endpoint
    endpoint = parsed.path
    if clean_query_params:
        query_string = "&".join([f"{k}={v[0]}" for k, v in clean_query_params.items()])
        endpoint = f"{endpoint}?{query_string}"
    
    connection_config = {
        "base_url": base_url,
        "endpoint": endpoint,
        "method": "GET",
        "prefer_http": ssl_config.get('prefer_http', False),
        "ssl_verify": ssl_config.get('ssl_verify', True)
    }
    
    credentials = {
        "api_key": api_key or "",
        "auth_header": auth_header
    }
    
    return connection_config, credentials


def validate_connection_url(url: str, connector_type: str) -> Tuple[bool, Optional[str]]:
    """
    Validate a connection URL without fully parsing it
    
    Returns:
        Tuple of (is_valid, error_message)
    """
    try:
        parse_connection_url(url, connector_type)
        return True, None
    except ConnectionParseError as e:
        return False, str(e)
    except Exception as e:
        return False, f"Unexpected error: {str(e)}"